### ISSUE TITLE
Proposal: Support retrieving application and request metadata

### ISSUE THREAD

**Issue by eandre** [encoredev/encore#171]
We've had several requests for more metadata about both the running Encore application and its environment, as well as about the currently running request. This proposal provides the suggested API for accomplishing these goals. Feedback is very much appreciated!

## Proposed API

```go
package encore // import "encore.dev"

// CurrentOp reports information about the current operation being handled
// by the calling goroutine. It never returns nil.
//
// It is safe to call concurrently from multiple goroutines and correctly handles
// the same server processing multiple concurrent requests.
func CurrentOp() *Operation

// Operation represents an incoming HTTP request or API call being handled by Encore.
type Operation struct {
	// Type is the type of operation this represents.
	Type OperationType
	// Service is the name of the service being called.
	Service string
	// Endpoint is the name of the endpoint being called.
	Endpoint string
	// Start is the time when the API endpoint was called to handle the request.
	Start time.Time
	// Path is the HTTP request path for the request.
	Path string
	// PathParams are the path parameters parsed from the request path.
	PathParams PathParams
	// Payload is the request payload parsed from the request body.
	// It is of the same type as the API endpoint's payload parameter.
	Payload any
}

type OperationType string

const (
    OpNone    OperationType = "none"
    OpAPICall OperationType = "api-call"
)

// PathParams contains the path parameters parsed from the request path.
type PathParams []PathParam

// PathParams represents a parsed path parameter.
type PathParam struct {
	Name  string // the name of the path parameter, without leading ':' or '*'.
	Value string // the parsed path parameter value.
}

// Get returns the value of the path parameter with the given name.
// If no such parameter exists it reports "".
func (p PathParams) Get(name string) string

// Meta reports metadata about the running Encore application.
// It never returns nil.
func Meta() *Metadata

// Metadata contains metadata about the running Encore application.
type Metadata struct {
	// AppID is the ID of the Encore application.
	// If the application is not linked to the Encore Platform it is "".
	AppID string

	// APIBaseURL is the base URL for calling this API.
	//
	// For local development it is "http://localhost:<port>", typically "http://localhost:4000".
	//
	// If a custom domain is used for this environment it is returned here, but note that
	// changes only take effect at the time of deployment while custom domains can be updated at any time.
	APIBaseURL string

	// EnvName is the name of the environment.
	// For local development it is "local".
	EnvName string

	// EnvType is the type of environment.
	EnvType EnvType

	// Cloud is the cloud provider this is running in.
	Cloud CloudProvider
}

// EnvType represents the type of environment.
// Additional environment types may be added in the future.
type EnvType string

const (
	// Production represents a production environment.
	Production EnvType = "production"
	// Development represents a long-lived cloud-hosted, non-production environment,
	// such as test environments.
	Development EnvType = "development"
	// Ephemeral represents short-lived cloud-hosted, non-production environments,
	// such as preview environments that only exist while a particular pull request is open.
	Ephemeral EnvType = "ephemeral"
	// Local represents the local development environment when using 'encore run'.
	Local EnvType = "local"
)

// CloudProvider represents the cloud provider this application is running in.
// Additional cloud providers may be added in the future.
type CloudProvider string

const (
	AWS CloudProvider = "aws"
	GCP CloudProvider = "gcp"
	Azure CloudProvider = "azure"
	EncoreCloud CloudProvider = "encorecloud"
	Local CloudProvider = "local"
)
```

**Comment by vilhelmmelkstam** [comment]
I really like being able to fetch these values and I think it is a big maturity step for encore! Also like the extensive number of values to fetch, good work!💪

The only question I had was regarding path parameters. What is the reason behind using a array of name/values instead of a map? Maybe it's just me who have gotten accustomed to it by using JS frameworks 😅

**Comment by eandre** [comment]
Thanks @vilhelmmelkstam!

> The only question I had was regarding path parameters. What is the reason behind using a array of name/values instead of a map? Maybe it's just me who have gotten accustomed to it by using JS frameworks 😅

That’s mainly for performance. Given the very very small number of path parameters in your typical endpoint it’s much faster to iterate over a slice than to do a hash map lookup.

**Comment by MatthewJamesBoyle** [comment]
Great proposal.  This would solve some issues I have been facing (I ended up using a secret to get the current environment) and grabbing path params has been something I struggled with.


Some suggestions:
```
func CurrentRequest() *Request
```
I think this should be 
```
func CurrentRequest() (Request,error)
```

For a couple of reasons:
- A pointer to the request would allow consumers to manipulate the request in unexpected ways which should be discouraged.
- Returning an error such as `errors.New("CurrentRequest cannot be called in an init func)` feels more idiomatic than it just being something callers need to know.

In the Request struct: 
```
Start time.Time
```
- wondering if this should be an epoch or a string in  RFC339 format so there is no confusion around timezones?
- Is is worth Request having an ID in the struct too? I know they have one anyway and so I see no harm in exposing it.
- Is it worth making Headers available from the Request too? 


All I can think of just now :)

**Comment by DomBlack** [comment]
Into the `Meta` data structure, I think we should include:
- The commit of the build.
- The rollout ID (while these two are closely linked they don't have to be the same)

In the `Request` data structure I think we should include the parent requests if one is present. Not sure how right now, but I'll think about it.

**Comment by DomBlack** [comment]
This was implemented in Encore v1.0
